### DISCUSSION TITLE
Fluentd image stops sending logs to Azure EventHub

### DISCUSSION THREAD

**Discussion by bilaljmi786** [community/community#134600] [Programming Help]
### Body

**2024-08-04 08:52:17.326224413 +0000 fluent.debug: {"message":"533 messages send."} -->** Messages sent correctly to Eventhub
**2024-08-04 08:52:18 +0000 [warn]: #0 fluent/log.rb:383:warn: Send exception occurred: Failed to send messages to application-logs/12**, --> Suddenly stops sending logs with this error

application-logs/26, application-logs/20

024-08-04 08:54:07 +0000 [debug]: #0 fluent/log.rb:341:debug: taking back chunk for errors. chunk="61ed7ad75c673bb242690f3353633346"
2024-08-04 08:54:07 +0000 [warn]: #0 fluent/log.rb:383:warn: **failed to flush the buffer**. retry_times=7 next_retry_time=2024-08-04 08:54:39 +0000 chunk="61ed7ad75c673bb242690f3353633346" **error_class=Kafka::DeliveryFailed error="Failed to send messages to** application-logs/19

@cosmo0920, any help would be appreciated as the application (fluentd) deployed in Azure Container Instance is crashing after a period of time. I have attached fluend logs from container for your reference.

The configuration of fluent.conf for buffer are as below,

	    # buffer settings
    	    <buffer topic>
        	@type file
        	path /var/log/td-agent/buffer/kafka
		flush_interval 2s
		flush_thread_count 2
   		retry_type exponential_backoff
   		retry_wait 1s
    		retry_max_interval 30s
   		retry_timeout 60m
   		chunk_limit_size 8m
   		total_limit_size 30g		
    	    </buffer>

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by cosmo0920** [comment]
CC: @ashie @daipom @kenhys

**Comment by geniuszly** [comment]
Hi! It seems like your Fluentd instance is encountering issues when sending logs to Azure EventHub, particularly related to the Kafka producer. The error logs indicate that after successfully sending a batch of messages, the application encounters a `Kafka::DeliveryFailed` error, which causes the buffer to be retried multiple times before ultimately failing.
1. Ensure that there is stable network connectivity between the Azure Container Instance (ACI) where Fluentd is running and the Azure EventHub. Any intermittent network issues could cause delivery failures.
2. Verify that the Azure Container Instance has sufficient resources (CPU, memory) to handle the load. If the container is running out of resources, it may cause Fluentd to crash or fail to send logs.
3. The current retry configuration might be insufficient for the load or network conditions. Consider increasing the `retry_max_interval` and `retry_timeout` to give more time for retries. For example:
```xml
retry_max_interval 60s
retry_timeout 120m
```
4. The buffer settings, such as `flush_interval`, `flush_thread_count`, and `chunk_limit_size`, might need adjustment. For high throughput scenarios, you may want to experiment with increasing `chunk_limit_size` and `flush_thread_count`. For example:
```xml
chunk_limit_size 16m
flush_thread_count 4
```
5. If you are using the Kafka plugin, ensure that the producer settings (e.g., `acks`, `compression_type`, `max_in_flight_requests_per_connection`) are optimized for your environment. For example:
```xml
acks "all"
compression_type "gzip"
max_in_flight_requests_per_connection 1
```
6. Ensure that your Azure EventHub is not hitting any quotas or throttling limits, which could impact message delivery.
7. Verify that the Kafka plugin version used in Fluentd is compatible with the Kafka version of your EventHub.
